subprocess_utils.hpp

```diff
@@ -0,0 +1,219 @@
+	     * */
+		void closeUnusedEnds() {
+			//we don't need the input end of the input pipe or the output end of the output pipe
+			close(input_pipe_file_descriptor[1]);
+			close(output_pipe_file_descriptor[0]);
+		}
+
+	public:
+	
+	    /**
+         * sets this to be the child end of the TwoWayPipe
+         * linking the input and output ends to stdin and stdout/stderr
+         * */
+		bool setAsChildEnd() {
+		    int tmp[2] = {input_pipe_file_descriptor[0],input_pipe_file_descriptor[1]};
+
+		    input_pipe_file_descriptor[0] = output_pipe_file_descriptor[0];
+			input_pipe_file_descriptor[1] = output_pipe_file_descriptor[1];
+			output_pipe_file_descriptor[0] = tmp[0];
+			output_pipe_file_descriptor[1] = tmp[1];
+		    
+			dup2(input_pipe_file_descriptor[0], STDIN_FILENO);
+			dup2(output_pipe_file_descriptor[1], STDOUT_FILENO);
+			dup2(output_pipe_file_descriptor[1], STDERR_FILENO);
+
+			closeUnusedEnds();
+		}
+        
+        /**
+         * sets this pipe to be the parent end of the TwoWayPipe
+         * */
+		bool setAsParentEnd() {
+			closeUnusedEnds();
+		}
+        
+        /**
+         * writes a string to the pipe 
+         * @param input - the string to write
+         * @return the number of bytes written
+         * */
+		size_t writeP(const std::string &input) {
+			return write(output_pipe_file_descriptor[1], input.c_str(), input.size());
+		}
+
+        /**
+         * @return true unless the last call to read either failed or reached EOF
+         * */
+        bool isGood(){
+            return inStreamGood;
+        }
+
+        /**
+    	* reads up to n bytes into the internal buffer
+    	* @param n - the max number of bytes to read in
+    	* @return the number of bytes read in, -1 in the case of an error
+    	* */
+        ssize_t readToInternalBuffer()
+        {
+            char buf[256];
+            int cnt;
+            ssize_t bytesCounted = -1;
+
+            while((bytesCounted = read(input_pipe_file_descriptor[0], buf, 256))<=0){
+            	if (bytesCounted < 0) {
+            	    if (errno != EINTR){ /* interrupted by sig handler return */
+            	        inStreamGood = false;
+                		return -1;
+            	    }
+            	}
+            	else if (bytesCounted == 0)  /* EOF */
+            	{
+            	    inStreamGood = false;
+            	    return 0;
+            	}
+            }
+            
+            internalBuffer.append(buf,bytesCounted);
+            return bytesCounted;
+        }
+
+		/**
+		* read line from the pipe - Not threadsafe
+		* blocks until either a newline is read or the other end of the pipe is closed
+		* @return the string read from the pipe or the empty string if there was not a line to read.
+		* */
+		std::string readLine() {
+		    size_t firstNewLine;
+		    size_t currentSearchPos = 0;
+		    while((firstNewLine = internalBuffer.find_first_of('\n',currentSearchPos))==std::string::npos)
```

Comment:
don't make me hit u 
```
foo() {

}
```

rather than
```
foo()
{

}
```

Reply:
also, be careful of spacing of binary operators - haven't looked too deeply, but that equality check is missing some padding spaces

Reply:
Yeah I'm having to do all the linux coding on an online emulator, not my usual environment. So I'm missing all the friendly automatic prettifying

Reply:
similar to what we do for souffle, I might set up a clang-tidy test, which makes sure that code is pretty otherwise it doesn't pass the test suite

Reply:
i shall ask martin on monday for details

Reply:
can't we make the tidy test beautify the code?

Reply:
tl;dr no you shouldn't

If the test beautifies the code, who should be attributed to commits?

We can setup git commit hooks which on your end performs them, but it requires all developers to install them (and if one forgets, then dirty code gets into the repo). The best procedure is to get developers to do it themselves, and then at the final stage, check whether its valid. You also need people to use the same *version* of clang-tidy, so it's best to have a central repo enforcing that.

Reply:
makes sense